VMware: return facts depending upon backing type

bushvin opened the pull request:
The class used for the backing is not identical in all cases, and do not have the same properties/methods.
Following backings exists at the time of this writing:
vim.vm.device.VirtualDisk.FlatVer1BackingInfo
vim.vm.device.VirtualDisk.FlatVer2BackingInfo
vim.vm.device.VirtualDisk.LocalPMemBackingInfo
vim.vm.device.VirtualDisk.PartitionedRawDiskVer2BackingInfo
vim.vm.device.VirtualDisk.RawDiskMappingVer1BackingInfo
vim.vm.device.VirtualDisk.RawDiskVer2BackingInfo
vim.vm.device.VirtualDisk.SeSparseBackingInfo
vim.vm.device.VirtualDisk.SparseVer1BackingInfo
vim.vm.device.VirtualDisk.SparseVer2BackingInfo

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

bushvin committed:
Not all Backings are equal

The class used for the backing is not identical in all cases, and do not have the same properties/methods.
Following backings exists at the time of this writing:
vim.vm.device.VirtualDisk.FlatVer1BackingInfo
vim.vm.device.VirtualDisk.FlatVer2BackingInfo
vim.vm.device.VirtualDisk.LocalPMemBackingInfo
vim.vm.device.VirtualDisk.PartitionedRawDiskVer2BackingInfo
vim.vm.device.VirtualDisk.RawDiskMappingVer1BackingInfo
vim.vm.device.VirtualDisk.RawDiskVer2BackingInfo
vim.vm.device.VirtualDisk.SeSparseBackingInfo
vim.vm.device.VirtualDisk.SparseVer1BackingInfo
vim.vm.device.VirtualDisk.SparseVer2BackingInfo

bushvin commented:
No idea whether it's either a Bugfix or Feature PR...

bushvin committed:
Per the request of my employer

Per the request of my employer. No other changes required.

bushvin commented:
Sorry about that last commit. They initially wanted nothing, but they changed thzir mind inbetween

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py:
```diff
@@ -147,16 +148,64 @@ def gather_disk_facts(self, vm_obj):
+                    disks_facts[disk_index]['backing_type'] = 'FlatVer1'
+                    disks_facts[disk_index]['backing_writethrough'] = disk.backing.writeThrough
+
+                elif isinstance(disk.backing, vim.vm.device.VirtualDisk.FlatVer2BackingInfo):
```
Please add example of these newly added keys in EXAMPLES section.

bushvin replied:
I updated the existing (FlatVer2) and added "RawDiskMappingVer1"

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py:
```diff
@@ -147,16 +148,64 @@ def gather_disk_facts(self, vm_obj):
+                elif isinstance(disk.backing, vim.vm.device.VirtualDisk.RawDiskMappingVer1BackingInfo):
+                    disks_facts[disk_index]['backing_type'] = 'RawDiskMappingVer1'
+                    disks_facts[disk_index]['backing_devicename'] = disk.backing.deviceName
+                    disks_facts[disk_index]['backing_diskmode'] = disk.backing.diskMode
```
Add a note in documentation saying - `Additional information about VM guest disk is added in version 2.8.`

bushvin replied:
Done

Akasurde commented:
@bushvin Thanks for the PR. Let me know once you are done with the requested changes. Thanks.

bushvin committed:
update

- Added update notification to notes
- Extended examples to match new return information (FlatVer2, RawDiskMappingVer1)

bushvin committed:
missing comma

Akasurde commented:
@pdellaert @dericcrago @Im0 @ckotte @jeking3 @Tomorrow9 @alongchamps @adarobin Could you please review this ? Thanks

jeking3 commented on lib/ansible/modules/cloud/vmware/vmware_guest_disk_facts.py:
```diff
@@ -147,16 +167,64 @@ def gather_disk_facts(self, vm_obj):
                     capacity_in_kb=disk.capacityInKB,
                     capacity_in_bytes=disk.capacityInBytes,
                 )
+                if isinstance(disk.backing, vim.vm.device.VirtualDisk.FlatVer1BackingInfo):
```
Just a commentary on the project's direction in general (not the PR author's) - having to transform results from the well-documented vSphere API to an arbitrary format per ansible module leads to chaos.

This style of rewriting results is unnecessary if you return the vSphere API object in json form... the vmware_guest_facts module in devel (and 2.8) can do this already, including returning complete details about the guest, including the disk backings, with all their details and without rewriting them.

Akasurde replied:
@jeking3 I totally agree with you. But before the change we didn't had an option to get facts about any VMware object from vCenter API. We can come up with strong proposal (not here of-course) and discuss in meeting so that we can arrive at roadmap kind of decision. Thanks for pointing this out.

alongchamps replied:
I'm in agreement that we should have a discussion on this. The more we keep in line with the VMware API, the easier it will be to traverse objects.

jeking3 reviewed (approved):
It would be nice to see the integration test (if there is one) updated to check the new field value for the backing type.

Akasurde commented:
> It would be nice to see the integration test (if there is one) updated to check the new field value for the backing type.

I tried govcsim with this module but it does not emulate different types of disk types.

Akasurde commented:
Closing and re-opening for CI trigger.
